### ISSUE TITLE
rspconfig allows get/set commands in the same string, allowing network config errors on typo

### ISSUE THREAD

**Issue by whowutwut** [xcat2/xcat-core#3383]
When running `rspconfig` found a few things that did not sit well.  

```
[root@fs3 ~]# rspconfig mid05tor12cn17 ip=50.6.31.1 netmask 255.0.0.0 gateway=50.2.5.3
mid05tor12cn17: BMC IP: 172.21.253.104
mid05tor12cn17: BMC Netmask: 255.255.0.0
mid05tor12cn17: BMC Gateway: 172.21.253.27
[root@fs3 ~]#
```

1) netmask doesn't have a = between netmask and 255.0.0.0 and the command did not error on this. So I assumed it was OKAY.  But really it did not set it.    I think we should check the arguments and all must be OK before we run any commands, After this is set, I'm no longer able to ping the BMC to change it becasue I moved it into another network. 

2) The output doesn't make sense, i'm setting the command but it's printing out the values that the BMC currently has

**Comment by xuweibj** [comment]
Hi @whowutwut , if the node is P9, setting is not supported even using ipmi now. 
For 1, my idea is to check arguments that do not allow setting and getting at the same time in one command. How do you think?
For 2, after setting will get the value from BMC. So the output will like that. We assume that the response for setting ip/gateway is OK from BMC using ipmi, and then try to get ip/gateway value and print BMC IP/Gateway as that.

**Comment by whowutwut** [comment]
@xuweibj  This is not run against a P9 machine...   It's run against a OpenPower P8 machine. 

1.  Is that what happened? It was displaying the values because I left out the `=`?    I don't think so because I think I ran this again afterwards and by default, it prints the existing value when I am setting values.. .

But yes, I think we should NOT allow setting and getting in the same command.  

2.  This is incorrect becasue we might be getting the OLD value before it's activated on the BMC.  If I'm setting the values, it would make more sense to print something like...
```
Attempting to set IP to X.X.X.X...
Attempting to set Netmask to X.X.X.X...
Attempting to set Gateway to X.X.X.X...
```

**Comment by xuweibj** [comment]
If it's run against P8, I think we need to check whether IP/Gateway is not correct value as setting.
For the command like that, the flow is as below:

``set BMC IP -> get BMC IP (If set successfully, the output will be the new value) -> get netmask -> set gateway -> get gateway (If set successfully, the output will be the new value)``

For ``255.0.0.0``, it is not valid subcommand, will not enter process function.

So I think we need to make sure whether set failed (ip/gateway) OR we get wrong value from BMC.

**Comment by whowutwut** [comment]
The command above in the original post is correct. The error is that I forgot to put an = between netmask and 255.0.0.0.  So the IP address and gateway was set.  But since the netmask was not set due to the typo. I was unable to communicate with the new IP address of the BMC due to my machine having a different 2nd octet. 

So my point is that if we have multiple arguments and it's not all set or all get. Flag an error or prevent the command from running because I can't reset after I lost the bmc. 

The output of the command is misleading because it prints the existing configuration... Not what I want to set.  So it's strange. Doesn't make any sense.

**Comment by xuweibj** [comment]
Hi @whowutwut , Ertao have tried this command on fs1. The output is: (The old netmask is 255.0.0.0)

```
64 bytes from 50.45.12.2: icmp_seq=34 ttl=64 time=60.9 ms
64 bytes from 50.45.12.2: icmp_seq=35 ttl=64 time=0.366 ms
^C
--- 50.45.12.2 ping statistics ---
35 packets transmitted, 18 received, +9 errors, 48% packet loss, time 34005ms
rtt min/avg/max/mdev = 0.356/57.243/450.952/134.868 ms, pipe 4
[root@fs1 ~]# chdef frame45cn12 bmc=50.45.12.2
1 object definitions have been created or modified.
[root@fs1 ~]# rspconfig frame45cn12 ip=50.45.12.1 netmask 255.0.0.0 gateway=50.3.29.1
frame45cn12: BMC IP: 50.45.12.1
frame45cn12: BMC Netmask: 255.0.0.0
frame45cn12: BMC Gateway: 50.3.29.1
[root@fs1 ~]# ping 50.45.12.1
PING 50.45.12.1 (50.45.12.1) 56(84) bytes of data.
From 50.3.29.1 icmp_seq=9 Destination Host Unreachable
From 50.3.29.1 icmp_seq=10 Destination Host Unreachable
From 50.3.29.1 icmp_seq=11 Destination Host Unreachable
From 50.3.29.1 icmp_seq=12 Destination Host Unreachable
From 50.3.29.1 icmp_seq=13 Destination Host Unreachable
From 50.3.29.1 icmp_seq=14 Destination Host Unreachable
64 bytes from 50.45.12.1: icmp_seq=15 ttl=64 time=1991 ms
64 bytes from 50.45.12.1: icmp_seq=16 ttl=64 time=990 ms
64 bytes from 50.45.12.1: icmp_seq=17 ttl=64 time=0.432 ms
64 bytes from 50.45.12.1: icmp_seq=18 ttl=64 time=442 ms
```

So if set successfully, will show new IP/Gateway. The result is from response of ipmi command. So I do know why in this issue you get the old value when you set successfully.

**Comment by whowutwut** [comment]
@xuweibj  I double checked it, and seems like the IP address did get set with the netmask.   So having a = or a space is OK?  

```
[root@c910f02c05p03 ~]# ipmitool -H 50.6.31.1 -U ADMIN -P ADMIN lan print 1
Set in Progress         : Set Complete
Auth Type Support       : NONE MD2 MD5 PASSWORD
Auth Type Enable        : Callback : MD2 MD5 PASSWORD
                        : User     : MD2 MD5 PASSWORD
                        : Operator : MD2 MD5 PASSWORD
                        : Admin    : MD2 MD5 PASSWORD
                        : OEM      : MD2 MD5 PASSWORD
IP Address Source       : Static Address
IP Address              : 50.6.31.1
Subnet Mask             : 255.0.0.0
MAC Address             : 0c:c4:7a:f4:5b:0e
SNMP Community String   : public
IP Header               : TTL=0x00 Flags=0x00 Precedence=0x00 TOS=0x00
BMC ARP Control         : ARP Responses Enabled, Gratuitous ARP Disabled
Default Gateway IP      : 50.2.5.3
Default Gateway MAC     : 7a:14:e3:95:b6:04
Backup Gateway IP       : 0.0.0.0
Backup Gateway MAC      : 00:00:00:00:00:00
802.1q VLAN ID          : Disabled
802.1q VLAN Priority    : 0
RMCP+ Cipher Suites     : 1,2,3,6,7,8,11,12
Cipher Suite Priv Max   : XaaaXXaaaXXaaXX
                        :     X=Cipher Suite Unused
                        :     c=CALLBACK
                        :     u=USER
                        :     o=OPERATOR
                        :     a=ADMIN
                        :     O=OEM
```

Maybe I did not wait long enough to try pinging this new IP address....

**Comment by whowutwut** [comment]
I wlll double check that and see if i can re-create it.

**Comment by whowutwut** [comment]
I spent some time here to try and understand what is going on.  @xuweibj  You are right, on fs1 the behavior seems to be correct.  I really can't explain why the output is what it is from fs3.  It seems that it printed the BMC information, instead of what we pass it from the command.  I can't explain that.  

So if the output is really what we set in the command line, then it's OK.  

Not allowing the get/set in the same line will help prevent the problem I hit... where I was changing the machine into another subnet.  The original subnet was netmask 255.255.0.0, the target subnet was 255.0.0.0.  The management node I was expecting to receive this node had an IP address of 50.3.X.X/8 and this target BMC is 50.6.X.X/16 .  I was trying to change it to /8 but because i forgot the `=`, it didn't take effect and not able to ping it....  Today I worked around it by creating an alias IP address on my target MGMT node with a 50.6.X.X/16 IP and now able to recover this node, but we should prevent this kind of typing mistake , which is what your pull request does.  

So long story short, I don't know how the fs3 output is what it is... but I'm pretty sure I did not just make it up, :-)   I always just copy/paste into the issues.

**Comment by whowutwut** [comment]
Alright, recreated.... 

```
[root@c910f02c05p03 ~]# lsxcatd -v
Version 2.13.6 (git commit 0e2c8c1d0d2b1ca2bee12f00dc678d8628c23602, built Thu Jul  6 06:15:37 EDT 2017)
```

Node definition:
```
[root@c910f02c05p03 ~]# lsdef boston01
Object name: boston01
    arch=ppc64le
    bmc=50.6.37.1
    bmcpassword=ADMIN
    bmcusername=ADMIN
    cons=ipmi
    groups=all,supermicro,boston
    installnic=mac
    ip=10.6.37.1
    mgt=ipmi
    netboot=petitboot
    postbootscripts=otherpkgs
    postscripts=syslog,remoteshell,syncfiles
    syslog=ignore
```

Then let's check the configuration:
```
[root@c910f02c05p03 ~]# rspconfig boston01 ip netmask gateway
boston01: BMC IP: 50.6.37.1
boston01: BMC Netmask: 255.255.0.0
boston01: BMC Gateway: 0.0.0.0
```

Now I want to fix this node for the netmask and gateway...
```
[root@c910f02c05p03 ~]# rspconfig boston01 netmask=255.0.0.0 gateway=50.2.5.3
boston01: BMC Netmask: 255.255.0.0
boston01: BMC Gateway: 0.0.0.0
```

So it prints out the netmask and gateway that is set on the BMC instead of the intended one the command line...

**Comment by xuweibj** [comment]
Hi @whowutwut , I added ipmitool command in code to check result, got result as below:

```
# rspconfig boston01 netmask=255.255.0.0 gateway=0.0.0.0
boston01: BMC Netmask: 255.0.0.0
boston01: Set in Progress         : Set Complete
Auth Type Support       : NONE MD2 MD5 PASSWORD
Auth Type Enable        : Callback : MD2 MD5 PASSWORD
                        : User     : MD2 MD5 PASSWORD
                        : Operator : MD2 MD5 PASSWORD
                        : Admin    : MD2 MD5 PASSWORD
                        : OEM      : MD2 MD5 PASSWORD
IP Address Source       : Static Address
IP Address              : 50.6.37.1
Subnet Mask             : 255.0.0.0
MAC Address             : 0c:c4:7a:f4:5b:07
SNMP Community String   : public
IP Header               : TTL=0x00 Flags=0x00 Precedence=0x00 TOS=0x00
BMC ARP Control         : ARP Responses Enabled, Gratuitous ARP Disabled
Default Gateway IP      : 50.2.5.3
Default Gateway MAC     : 7a:14:e3:95:b6:04
Backup Gateway IP       : 0.0.0.0
Backup Gateway MAC      : 00:00:00:00:00:00
802.1q VLAN ID          : Disabled
802.1q VLAN Priority    : 0
RMCP+ Cipher Suites     : 1,2,3,6,7,8,11,12
Cipher Suite Priv Max   : XaaaXXaaaXXaaXX
                        :     X=Cipher Suite Unused
                        :     c=CALLBACK
                        :     u=USER
                        :     o=OPERATOR
                        :     a=ADMIN
                        :     O=OEM
Bad Password Threshold  : 5
Invalid password disable: yes
Attempt Count Reset Int.: 600
User Lockout Interval   : 600

boston01: BMC Gateway: 50.2.5.3
boston01: Set in Progress         : Set Complete
Auth Type Support       : NONE MD2 MD5 PASSWORD
Auth Type Enable        : Callback : MD2 MD5 PASSWORD
                        : User     : MD2 MD5 PASSWORD
                        : Operator : MD2 MD5 PASSWORD
                        : Admin    : MD2 MD5 PASSWORD
                        : OEM      : MD2 MD5 PASSWORD
IP Address Source       : Static Address
IP Address              : 50.6.37.1
Subnet Mask             : 255.0.0.0
MAC Address             : 0c:c4:7a:f4:5b:07
SNMP Community String   : public
IP Header               : TTL=0x00 Flags=0x00 Precedence=0x00 TOS=0x00
BMC ARP Control         : ARP Responses Enabled, Gratuitous ARP Disabled
Default Gateway IP      : 50.2.5.3
Default Gateway MAC     : 7a:14:e3:95:b6:04
Backup Gateway IP       : 0.0.0.0
Backup Gateway MAC      : 00:00:00:00:00:00
802.1q VLAN ID          : Disabled
802.1q VLAN Priority    : 0
RMCP+ Cipher Suites     : 1,2,3,6,7,8,11,12
Cipher Suite Priv Max   : XaaaXXaaaXXaaXX
                        :     X=Cipher Suite Unused
                        :     c=CALLBACK
                        :     u=USER
                        :     o=OPERATOR
                        :     a=ADMIN
                        :     O=OEM
Bad Password Threshold  : 5
Invalid password disable: yes
Attempt Count Reset Int.: 600
User Lockout Interval   : 600
```

Just as you got, **Subnet Mask             : 255.0.0.0**, **Default Gateway IP      : 50.2.5.3**, though we set it successfully we got old value from BMC even use ipmitool command. It seems that this BMC has some problem.

**Comment by whowutwut** [comment]
@xuweibj Thanks for following up, so it looks like this is an issue with the SMC servers...  or a support issue since we have not validated this ...  

@zet809 Comparing Boston with Firestone, it looks like the IPMI commands are the same...  
```
[root@c910f02c05p03 ~]# rspconfig boston01 netmask=255.0.0.0 gateway=50.2.5.3
[ipmi_debug] rspconfig: netmask=255.0.0.0(netmask=255.0.0.0 gateway=50.2.5.3), raw_cmd: netfn(0x06=>App), cmd(0x38=>Get Channel Authentication Capabilities), data=[142, 4]
[ipmi_debug] rspconfig: netmask=255.0.0.0(netmask=255.0.0.0 gateway=50.2.5.3), raw_cmd: netfn(0x06=>App), cmd(0x3b=>Set session privilege level), data=[4]
[ipmi_debug] rspconfig: netmask=255.0.0.0(gateway=50.2.5.3), raw_cmd: netfn(0x0c=>Transport), cmd(0x01=>Set LAN Configuration Parameters), data=[1, 0, 1]
[ipmi_debug] rspconfig: netmask=255.0.0.0(gateway=50.2.5.3), raw_cmd: netfn(0x0c=>Transport), cmd(0x01=>Set LAN Configuration Parameters), data=[1, 6, 255, 0, 0, 0]
[ipmi_debug] rspconfig: netmask=255.0.0.0(gateway=50.2.5.3), raw_cmd: netfn(0x0c=>Transport), cmd(0x01=>Set LAN Configuration Parameters), data=[1, 0, 0]
[ipmi_debug] rspconfig: netmask=255.0.0.0(gateway=50.2.5.3), raw_cmd: netfn(0x0c=>Transport), cmd(0x02=>Get LAN Configuration Parameters), data=[1, 6, 0, 0]
boston01: BMC Netmask: 255.255.0.0
[ipmi_debug] rspconfig: gateway=50.2.5.3(), raw_cmd: netfn(0x0c=>Transport), cmd(0x01=>Set LAN Configuration Parameters), data=[1, 0, 1]
[ipmi_debug] rspconfig: gateway=50.2.5.3(), raw_cmd: netfn(0x0c=>Transport), cmd(0x01=>Set LAN Configuration Parameters), data=[1, 12, 50, 2, 5, 3]
[ipmi_debug] rspconfig: gateway=50.2.5.3(), raw_cmd: netfn(0x0c=>Transport), cmd(0x01=>Set LAN Configuration Parameters), data=[1, 0, 0]
[ipmi_debug] rspconfig: gateway=50.2.5.3(), raw_cmd: netfn(0x0c=>Transport), cmd(0x02=>Get LAN Configuration Parameters), data=[1, 12, 0, 0]
boston01: BMC Gateway: 0.0.0.0
[ipmi_debug] rspconfig: (), raw_cmd: netfn(0x0c=>Transport), cmd(0x02=>Get LAN Configuration Parameters), data=[1, 12, 0, 0]
[ipmi_debug] rspconfig: (), raw_cmd: netfn(0x0c=>Transport), cmd(0x02=>Get LAN Configuration Parameters), data=[1, 6, 0, 0]
[ipmi_debug] rspconfig: (), raw_cmd: netfn(0x06=>App), cmd(0x3c=>Close Session), data=[195, 0, 0, 0]
```

```
[root@fs1 ~]# rspconfig frame45cn12 ip=50.45.12.2 netmask 255.255.0.0 gateway
[ipmi_debug] rspconfig: ip=50.45.12.2(ip=50.45.12.2 netmask 255.255.0.0 gateway), raw_cmd: netfn(0x06=>App), cmd(0x38=>Get Channel Authentication Capabilities), data=[142, 4]
[ipmi_debug] rspconfig: ip=50.45.12.2(ip=50.45.12.2 netmask 255.255.0.0 gateway), raw_cmd: netfn(0x06=>App), cmd(0x3b=>Set session privilege level), data=[4]
[ipmi_debug] rspconfig: ip=50.45.12.2(netmask 255.255.0.0 gateway), raw_cmd: netfn(0x0c=>Transport), cmd(0x01=>Set LAN Configuration Parameters), data=[1, 0, 1]
[ipmi_debug] rspconfig: ip=50.45.12.2(netmask 255.255.0.0 gateway), raw_cmd: netfn(0x0c=>Transport), cmd(0x01=>Set LAN Configuration Parameters), data=[1, 4, 1]
[ipmi_debug] rspconfig: ip=50.45.12.2(netmask 255.255.0.0 gateway), raw_cmd: netfn(0x0c=>Transport), cmd(0x01=>Set LAN Configuration Parameters), data=[1, 3, 50, 45, 12, 2]
[ipmi_debug] rspconfig: ip=50.45.12.2(netmask 255.255.0.0 gateway), raw_cmd: netfn(0x0c=>Transport), cmd(0x01=>Set LAN Configuration Parameters), data=[1, 0, 0]
[ipmi_debug] rspconfig: ip=50.45.12.2(netmask 255.255.0.0 gateway), raw_cmd: netfn(0x0c=>Transport), cmd(0x02=>Get LAN Configuration Parameters), data=[1, 3, 0, 0]
frame45cn12: BMC IP: 50.45.12.2
[ipmi_debug] rspconfig: netmask(255.255.0.0 gateway), raw_cmd: netfn(0x0c=>Transport), cmd(0x02=>Get LAN Configuration Parameters), data=[1, 6, 0, 0]
frame45cn12: BMC Netmask: 255.0.0.0
[ipmi_debug] rspconfig: 255.255.0.0(gateway), raw_cmd: netfn(0x0c=>Transport), cmd(0x02=>Get LAN Configuration Parameters), data=[1, 3, 0, 0]
[ipmi_debug] rspconfig: gateway(), raw_cmd: netfn(0x0c=>Transport), cmd(0x02=>Get LAN Configuration Parameters), data=[1, 12, 0, 0]
frame45cn12: BMC Gateway: 50.3.29.1
[ipmi_debug] rspconfig: (), raw_cmd: netfn(0x06=>App), cmd(0x3c=>Close Session), data=[80, 103, 250, 19]
```

We do a SET LAN before we do  a GET LAN configuration.  Perhaps there is a time delay between when the SET is called and when the GET will reflect the new value??

**Comment by whowutwut** [comment]
Opened new issue to track the feature support, using this issue to handle the change to not allow GET/SET in the same line.

**Comment by whowutwut** [comment]
Verified on 7/17 daily build. 

OK to close.  

```
[root@fs3 ~]# rspconfig mid05tor12cn07 ip netmask=255.255.0.0 gateway
Usage Error: Cannot display and change attributes on the same command.
```
